### PR TITLE
PLFM-4590 JSON Web Token component

### PR DESCRIPTION
**Pull request by brucehoff**
We use signed JSON Web Tokens (JWTs) to encode OIDC user ID tokens and also OAuth Access Tokens.
The tokens are signed using asymmetric encryption.
The public key(s) is(are) published as a JSON Web Key Set.
The body of a JWT is a JSON object.  Some entries are predefined token metadata.
The body of an access token also contains information (scopes and claims) about what the bearer is allowed to access.
The body of an ID token contains information (claims) about the user (the "subject").
The Docker registry integration already makes use of JWTs for authorizing the Docker client.  This PR refactors the common code so it can be used by both the Docker and OIDC code.

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/KeyPairUtil.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	 * so this will not work for non-RSA keys.
+	 * 
+	 * @param pemEncodedPrivateKey
+	 * @return
+	 */
+	public static KeyPair getRSAKeyPairFromPrivateKey(String pemEncodedPrivateKey) {
+		try {
+			byte[] content = Base64.decodeBase64(pemEncodedPrivateKey);
+			PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(content);
+			KeyFactory factory = KeyFactory.getInstance("RSA");
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/KeyPairUtil.java:55]: define a constant for "RSA" (or use an existing one)

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/KeyPairUtil.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		}
+	}
+	
+	// from https://botbot.me/freenode/cryptography-dev/2015-12-04/?page=1
+	// SPKI DER SHA-256 hash, strip of the last two bytes, base32 encode it and then add a : every four chars.
+	public static String computeKeyId(PublicKey publicKey) {
+		try {
+			if (publicKey==null) throw new RuntimeException();
+			// http://stackoverflow.com/questions/3103652/hash-string-via-sha-256-in-java
+			MessageDigest md = MessageDigest.getInstance("SHA-256");
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/KeyPairUtil.java:74]: define constant for SHA-256 (or use existing one)

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	private static final String SCOPE = "scope";
+	private static final String USER_INFO_CLAIMS = "oidc_claims";
+	private static final String NULL = "null";
+	
+	/*
+	 * Given lists of scopes and claims, build an 'access' claim JSONObject and add it to the 
+	 * 'claims' object of a JWT-encoded access token
+	 */
+	public static void addAccessClaims(List<OAuthScope> scopes, Map<OIDCClaimName, OIDCClaimsRequestDetails> oidcClaims, Claims claims) {
+		Map<String,Object> scopeAndClaims = new HashMap<String,Object>();
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java:32]: Use immutable map (Guava library).  Of static method creates it.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		scopeAndClaims.put(USER_INFO_CLAIMS, oidcClaims);		
+		claims.put(ACCESS, scopeAndClaims);
+	}
+
+	/* 
+	 * Given the claims from a JWT-encoded access token, extract the list of granted scopes.
+	 */
+	public static List<OAuthScope> getScopeFromClaims(Claims claims) {
+		Map<String,Object> scopeAndClaims = (Map<String,Object>)claims.get(ACCESS, Map.class);
+		return (List<OAuthScope>)scopeAndClaims.get(SCOPE);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java:43]: check for scopeAndClaims==null, return Collections.emptyList()

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	
+	/* 
+	 * Given the claims from a JWT-encoded access token, extract the list of granted claims.
+	 */
+	public static Map<OIDCClaimName, OIDCClaimsRequestDetails> getOIDCClaimsFromClaimSet(Claims claims) {
+		Map<String,Object> scopeAndClaims = (Map<String,Object>)claims.get(ACCESS, Map.class);
+		return (Map<OIDCClaimName, OIDCClaimsRequestDetails>)scopeAndClaims.get(USER_INFO_CLAIMS);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java:51]: check for scopeAndClaims==null, return Collections.emptyList()

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	 */
+	public static Map<OIDCClaimName, OIDCClaimsRequestDetails> getClaimsMapFromJSONObject(JSONObject claimsJson) {
+		Map<OIDCClaimName, OIDCClaimsRequestDetails> result = new HashMap<OIDCClaimName, OIDCClaimsRequestDetails>();
+		if (claimsJson==null) return result;
+		for (String claimName : claimsJson.keySet()) {
+			OIDCClaimName claim;
+			try {
+				claim = OIDCClaimName.valueOf(claimName);
+			} catch (IllegalArgumentException e) {
+				continue; // ignore unknown claims
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java:68]: link to the spec' that says unrecognized claims are ignored

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+			}
+			String detailsString = claimsJson.getString(claimName);
+			OIDCClaimsRequestDetails details = null;
+			if (detailsString!=null && !NULL.equals(detailsString)) {
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtil.java:72]: equalsIgnoreCase

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		JSONObject claimsJson = new JSONObject();
+		claimsJson.put(OIDCClaimName.team.name(), "{\"values\":[\"101\"]}");
+		claimsJson.put(OIDCClaimName.given_name.name(), "null");
+		claimsJson.put(OIDCClaimName.family_name.name(), "{\"essential\":true,\"value\":\"foo\"}");
+		claimsJson.put("unrecognized-key", "some value");
+		
+		// method under test
+		Map<OIDCClaimName, OIDCClaimsRequestDetails> actual = ClaimsJsonUtil.getClaimsMapFromJSONObject(claimsJson);
+		
+		OIDCClaimsRequestDetails expected = new OIDCClaimsRequestDetails();
```
Comment (brucehoff) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtilTest.java:35]: expectedTeam, expectedGiveName, expectedFamilyName

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jwts;
+
+public class ClaimsJsonUtilTest {
+
+	@Test
+	public void testGetClaimsMapFromJSONObject() {
+		JSONObject claimsJson = new JSONObject();
+		claimsJson.put(OIDCClaimName.team.name(), "{\"values\":[\"101\"]}");
+		claimsJson.put(OIDCClaimName.given_name.name(), "null");
```
Comment (brucehoff) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/ClaimsJsonUtilTest.java:28]: upper case NULL

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelper.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	 * @param authTimeSeconds The timestamp for the event in which the user most recently logged in to Synapse
+	 * @param tokenId a unique ID for this token
+	 * @param scopes the authorized scopes.  To retrieve user info' claims, the 'openid' scope is required
+	 * @param oidcClaims the fine-grained details about what user info can be accessed by this access token
+	 * @return a serialized JSON Web Token
+	 */
+	String createOIDCaccessToken(String issuer, String subject, String oauthClientId, long now, Long authTimeSeconds,
+			String tokenId, List<OAuthScope> scopes, Map<OIDCClaimName, OIDCClaimsRequestDetails> oidcClaims);
+
+	JsonWebKeySet getJSONWebKeySet();
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelper.java:49]: comment method signature

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelper.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+	/**
+	 * Validate the given JWT.  If valid, return the content as a Jwt object otherwise return null
+	 * @param a serialized JSON Web Token
+	 * @return the parsed and validated JWT or null if not valid
+	 */
+	Jwt<JwsHeader,Claims> validateJWTSignature(String token);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelper.java:56]: expose two methods: (1) validateJWT which throws exception if invalid, (2) parseJWT which returns parsed token

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+	@Autowired
+	private StackConfiguration stackConfiguration;
+	
+	@Override
+	public void afterPropertiesSet() {
+		jsonWebKeySet = new JsonWebKeySet();
+		List<JsonWebKey> publicKeys = new ArrayList<JsonWebKey>();
+		jsonWebKeySet.setKeys(publicKeys);
+		for (String s: stackConfiguration.getOIDCSignatureRSAPrivateKeys()) {
+			KeyPair keyPair = KeyPairUtil.getRSAKeyPairFromPrivateKey(s);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java:56]: put logic into a utility method (and write a test)

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+		return createSignedJWT(claims);
+	}
+	
+	public static RSAPublicKey getRSAPublicKeyForJsonWebKeyRSA(JsonWebKeyRSA jwkRsa) {
+		BigInteger modulus = new BigInteger(jwkRsa.getN());
+		BigInteger publicExponent = new BigInteger(jwkRsa.getE());
+		RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, publicExponent);
+		try {
+			KeyFactory kf = KeyFactory.getInstance("RSA");
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java:152]: use a static string for "RSA".  Put in the KeyPairUtil.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+		// find the key Id, then, once we map the key Id to the signing key,
+		// we parse again, setting the matching public key for verification
+		String[] pieces = token.split("\\.");
+		if (pieces.length!=3) throw new IllegalArgumentException("Expected three pieces but found "+pieces.length);
+		String unsignedToken = pieces[0]+"."+pieces[1]+".";
+		JsonWebKey matchingKey=null;
+		{
+			Jwt<Header,Claims> unsignedJwt = Jwts.parser().parseClaimsJwt(unsignedToken);
+			for (JsonWebKey jwk : jsonWebKeySet.getKeys()) {
+				if (jwk.getKid().equals(unsignedJwt.getHeader().get("kid"))) {
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java:171]: "kid" -> string constant

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImplTest.java [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+		Claims claims = jwt.getBody();
+		// here we just check that the 'access' claim has been added
+		// in the test for ClaimsJsonUtil.addAccessClaims() we check 
+		// that the content is correct
+		assertNotNull(claims.get("access"));
+		
+		// This checks the other fields set in the method under test
+		clientValidation(accessToken, null/* no nonce */);
+		
+		System.out.println(claims);
```
Comment (brucehoff) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImplTest.java:233]: remove print statement

**Review by marcomarasca — APPROVED** [review]

**Pull request merged**